odoo/models.py

```diff
@@ -5532,27 +5523,35 @@ def blacklist_given_fields(model):
                     blacklist.update(set(self.env[parent_model]._fields) - whitelist)
                 else:
                     blacklist_given_fields(self.env[parent_model])
-
         blacklist_given_fields(self)
 
         fields_to_copy = {name: field
                           for name, field in self._fields.items()
                           if field.copy and name not in default and name not in blacklist}
 
-        for name, field in fields_to_copy.items():
-            if field.type == 'one2many':
-                # duplicate following the order of the ids because we'll rely on
-                # it later for copying translations in copy_translation()!
-                lines = [rec.copy_data()[0] for rec in self[name].sorted(key='id')]
-                # the lines are duplicated using the wrong (old) parent, but then are
-                # reassigned to the correct one thanks to the (Command.CREATE, 0, ...)
-                default[name] = [Command.create(line) for line in lines if line]
-            elif field.type == 'many2many':
-                default[name] = [Command.set(self[name].ids)]
-            else:
-                default[name] = field.convert_to_write(self[name], self)
-
-        return [default]
+        for record in self:
+            seen_map = self._context['__copy_data_seen']
+            if record.id in seen_map[record._name]:
+                result.append(None)
+                continue
+            seen_map[record._name].add(record.id)
+
+            record_default = default.copy()
```

Comment:
Using the name "default" in the original code was misleading.
I suggest to rename `result` to `vals_list` and `record_default` to `vals`.